{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "969740fa",
   "metadata": {},
   "source": [
    "# Getting familiar with LIANA\n",
    "\n",
    "`liana` provides different statistical methods to infer `ligand-receptor` interactions from single-cell transcriptomics data omics data using prior knowledge.\n",
    "In this notebook we showcase how to use liana in its most basic form with toy data.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe2b9d28",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import liana as li"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "550a56c1",
   "metadata": {},
   "source": [
    "## load example dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7c1372e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.datasets.pbmc68k_reduced()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4a9da8f",
   "metadata": {},
   "source": [
    "`liana` typically works with the log1p-trasformed counts matrix, in this object the normalized counts are stored in `raw`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "777d5842",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.raw.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c5cc873",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color='bulk_labels', title='', frameon=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6149e156",
   "metadata": {},
   "source": [
    "## Methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fae5ce2",
   "metadata": {},
   "outputs": [],
   "source": [
    "li.mt.show_methods()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9386c398",
   "metadata": {},
   "source": [
    "Each method infers relevant ligand-receptor interactions relying on different assumptions and each method returns different ligand-receptor scores, typically a pair per method. One score corresponding to\n",
    "the `magnitude` (strength) of interaction and the other reflecting how `specificity` of a given interaction to a pair cell identities."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a70ee24",
   "metadata": {},
   "source": [
    "## Resources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d81d2033",
   "metadata": {},
   "outputs": [],
   "source": [
    "li.rs.show_resources()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97d7ef60",
   "metadata": {},
   "source": [
    "By default, liana uses the consensus resource, which is composed by multiple expert-curated ligand-receptor resources, including CellPhoneDB, CellChat, ICELLNET, connectomeDB2020, and CellTalkDB."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8d0d05b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from liana.mt import cellchat\n",
    "?cellchat.__call__"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a945dcd",
   "metadata": {},
   "source": [
    "## combine the scores of multiple methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d42c90d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from liana.mt import rank_aggregate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fabbcde1",
   "metadata": {},
   "outputs": [],
   "source": [
    "rank_aggregate.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfec0f75",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import all individual methods\n",
    "from liana.method import singlecellsignalr, connectome, cellphonedb, natmi, logfc, cellchat, geometric_mean"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53a43cc3",
   "metadata": {},
   "source": [
    "## run cellphonedb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67b5077d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cellphonedb(adata,\n",
    "            groupby='bulk_labels', \n",
    "            # NOTE by default the resource uses HUMAN gene symbols\n",
    "            resource_name='consensus',\n",
    "            expr_prop=0.1,\n",
    "            verbose=True, key_added='cpdb_res')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fef3aa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# by default, liana's output is saved in place:\n",
    "adata.uns['cpdb_res'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dfe4e49",
   "metadata": {},
   "source": [
    "### Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69fe70c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "li.pl.dotplot(adata = adata, \n",
    "              colour='lr_means',\n",
    "              size='cellphone_pvals',\n",
    "              inverse_size=True, # we inverse sign since we want small p-values to have large sizes\n",
    "              source_labels=['CD34+', 'CD56+ NK', 'CD14+ Monocyte'],\n",
    "              target_labels=['CD34+', 'CD56+ NK'],\n",
    "              figure_size=(8, 7),\n",
    "              # finally, since cpdbv2 suggests using a filter to FPs\n",
    "              # we filter the pvals column to <= 0.05\n",
    "              filter_fun=lambda x: x['cellphone_pvals'] <= 0.05,\n",
    "              uns_key='cpdb_res' # uns_key to use, default is 'liana_res' \n",
    "             )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95515ba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_plot = li.pl.tileplot(adata = adata, \n",
    "                         # NOTE: fill & label need to exist for both\n",
    "                         # ligand_ and receptor_ columns\n",
    "                         fill='means',\n",
    "                         label='props',\n",
    "                         label_fun=lambda x: f'{x:.2f}',\n",
    "                         top_n=10, \n",
    "                         orderby='cellphone_pvals',\n",
    "                         orderby_ascending=True,\n",
    "                         source_labels=['CD34+', 'CD56+ NK', 'CD14+ Monocyte'],\n",
    "                         target_labels=['CD34+', 'CD56+ NK'],\n",
    "                         uns_key='cpdb_res', # NOTE: default is 'liana_res'\n",
    "                         source_title='Ligand',\n",
    "                         target_title='Receptor',\n",
    "                         figure_size=(8, 7)\n",
    "                         )\n",
    "my_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee1b755b",
   "metadata": {},
   "source": [
    "## Ranking and aggregating"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "605ad8a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run rank_aggregate\n",
    "li.mt.rank_aggregate(adata, \n",
    "                     groupby='bulk_labels',\n",
    "                     resource_name='consensus',\n",
    "                     expr_prop=0.1,\n",
    "                     verbose=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff9a411b",
   "metadata": {},
   "outputs": [],
   "source": [
    "?li.mt.rank_aggregate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "228b4ffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns['liana_res'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab701035",
   "metadata": {},
   "source": [
    "### Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26ef8595",
   "metadata": {},
   "outputs": [],
   "source": [
    "li.pl.dotplot(adata = adata, \n",
    "              colour='magnitude_rank',\n",
    "              size='specificity_rank',\n",
    "              inverse_size=True,\n",
    "              inverse_colour=True,\n",
    "              source_labels=['CD34+', 'CD56+ NK', 'CD14+ Monocyte'],\n",
    "              target_labels=['CD34+', 'CD56+ NK'],\n",
    "              top_n=10, \n",
    "              orderby='magnitude_rank',\n",
    "              orderby_ascending=True,\n",
    "              figure_size=(9, 5)\n",
    "             )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f7789b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_plot = li.pl.dotplot(adata = adata, \n",
    "                        colour='magnitude_rank',\n",
    "                        inverse_colour=True,\n",
    "                        size='specificity_rank',\n",
    "                        inverse_size=True,\n",
    "                        source_labels=['CD34+', 'CD56+ NK', 'CD14+ Monocyte'],\n",
    "                        target_labels=['CD34+', 'CD56+ NK'],\n",
    "                        filter_fun=lambda x: x['specificity_rank'] <= 0.01,\n",
    "                       )\n",
    "my_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b6d9d5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save the plot to a file\n",
    "my_plot.save('dotplot.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27ad1b19",
   "metadata": {},
   "source": [
    "## Customizing LIANA's rank aggregate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c44419d",
   "metadata": {},
   "outputs": [],
   "source": [
    "li.mt.show_methods()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fb43016",
   "metadata": {},
   "outputs": [],
   "source": [
    "# connectome, cellphonedb, natmi, logfc, cellchat, geometric_mean\n",
    "methods = [cellchat, logfc,connectome]\n",
    "new_rank_aggregate = li.mt.AggregateClass(li.mt.aggregate_meta, methods=methods)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cbc0d34",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_rank_aggregate(adata,\n",
    "                   groupby='bulk_labels',\n",
    "                   expr_prop=0.1, \n",
    "                   verbose=True,\n",
    "                   n_perms=500,\n",
    "                   #n_perms=None, # you can also don't perform permutations to increase the speed, and exclude p-values and specificity_rank\n",
    "                   use_raw=True,\n",
    "                   )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d8ba4db",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns['liana_res'].head() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d1f4fe2",
   "metadata": {},
   "outputs": [],
   "source": [
    "li.pl.dotplot(adata = adata, \n",
    "              colour='magnitude_rank',\n",
    "              size='specificity_rank',\n",
    "              inverse_size=True,\n",
    "              inverse_colour=True,\n",
    "              source_labels=['CD34+', 'CD56+ NK', 'CD14+ Monocyte'],\n",
    "              target_labels=['CD34+', 'CD56+ NK', 'CD14+ Monocyte'],\n",
    "              top_n=10, \n",
    "              orderby='magnitude_rank',\n",
    "              orderby_ascending=True,\n",
    "              figure_size=(9, 5)\n",
    "             )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ed131a0",
   "metadata": {},
   "source": [
    "### Customizing LIANA's Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daad78e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotnine as p9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db54e3ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "(my_plot +\n",
    " # change theme\n",
    " p9.theme_dark() +\n",
    " # modify theme\n",
    " p9.theme(\n",
    "     # adjust facet size\n",
    "     strip_text=p9.element_text(size=11),\n",
    "     figure_size=(7, 4)\n",
    " )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da88d80f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Circle Plot\n",
    "li.pl.circle_plot(adata,\n",
    "                  groupby='bulk_labels',\n",
    "                  score_key='magnitude_rank',\n",
    "                  inverse_score=True,\n",
    "                  source_labels='CD34+',\n",
    "                 # filter_fun=lambda x: x['specificity_rank'] <= 0.05,\n",
    "                  pivot_mode='counts', # NOTE: this will simply count the interactions, 'mean' is also available\n",
    "                  figure_size=(10, 10),\n",
    "                  )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fb4f221",
   "metadata": {},
   "source": [
    "# Practice"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06a5d8a0",
   "metadata": {},
   "source": [
    "Use the use an ischemic 10X Visium spatial slide from [Kuppe et al., 2022] `kuppe_heart19.h5ad` dataset to practice:\n",
    "- calculating ligand-receptor interactions per cell type using rank_aggregate method;\n",
    "- visualize the top 10 ligand-receptor interactions of each cell type;\n",
    "\n",
    "Tips: \n",
    "- The cell-type compositions per spot are stored in obsm['compositions']. You can assign each spot to the cell type with the highest value in 'compositions' by setting adata.obs['cell_type'].\n",
    "- Don't forget to normalize the data before running LIANA."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (02 Liana)",
   "language": "python",
   "name": "02-liana"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
