{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Learning Spatial Relationships with MISTy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While Moran's R provides a sound summary of spatial clustering, it is limited to two variables at a time and is thus not fit for **complex, or non-linear, spatial relationships** between variables. \n",
    "\n",
    "Here, we show how to use LIANA's implementation of [MISTy](https://github.com/saezlab/mistyR), a framework presented in [Tanevski et al., 2022](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02663-5).\n",
    "\n",
    "**MISTy** is a tool that helps us better understand how different features, such as genes or cell types, interact with each other in space. MISTy does so by learning both **intra-** and **extracellular** relationships - i.e. those that occur within and between cells/spots. **A major advantage of MISTy is its flexibility**. It can model different perspectives, or \"views,\" each describing a different way markers are related to each other. Each of these views can describe a different spatial context, i.e. define a relationship among the observed expressions of the markers, such as intracellular regulation or paracrine regulation.\n",
    "\n",
    "**MISTy has only one fixed view** - i.e. the **intraview**, which contains the target (dependent) variables. The other views we refer to as extra views, and they contain the independent variables used to predict the intra view. **MISTy can fit any number of extra views, and each extra view can contain any number of variables.** The extra views can thus simultaneously **learn the dependencies of target variables across different modalities**, such as cell type proportions, pathways, or genes, etc.\n",
    "\n",
    "MISTy represents each view represents as a potential source of variation in the measurements of the target  variables in the intra view. MISTy further analyzes each view to determine how it contributes to the overall expression or abundance of each target variable. It explains this contribution by identifying the interactions between measurements that led to the observed results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import decoupler as dc\n",
    "import plotnine as p9\n",
    "import liana as li\n",
    "import os\n",
    "\n",
    "datadir = '../../datasets/Hands_on_2_LIANA_MistY/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import functions needed to create MISTy objects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from liana.method import MistyData, genericMistyData, lrMistyData"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import Pre-defined Single view models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from liana.method.sp import RandomForestModel, LinearModel, RobustLinearModel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load and Normalize Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We still use \"kuppe_heart19.h5ad\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read(os.path.join(datadir, \"kuppe_heart19.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.layers['counts'] = adata.X.copy()\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Extract Cell type Composition\n",
    "This slide comes with estimated cell type proportions using cell2location; See [Kuppe et al., 2022](https://www.nature.com/articles/s41586-022-05060-x). Let's extract from .obsm them to an independent AnnData object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename to more informative names\n",
    "full_names = {'Adipo': 'Adipocytes',\n",
    "              'CM': 'Cardiomyocytes',\n",
    "              'Endo': 'Endothelial',\n",
    "              'Fib': 'Fibroblasts',\n",
    "              'PC': 'Pericytes',\n",
    "              'prolif': 'Proliferating',\n",
    "              'vSMCs': 'Vascular_SMCs',\n",
    "              }\n",
    "# but only for the ones that are in the data\n",
    "adata.obsm['compositions'].columns = [full_names.get(c, c) for c in adata.obsm['compositions'].columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "comps = li.ut.obsm_to_adata(adata, 'compositions')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "comps.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check key cell types\n",
    "sc.pl.spatial(comps,\n",
    "              color=['Vascular_SMCs','Cardiomyocytes',\n",
    "                     'Endothelial', 'Fibroblasts'],\n",
    "              size=1.3, ncols=2, alpha_img=0\n",
    "              )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Formatting & Running MISTy\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The implementation of MISTy in LIANA relies on [MuData](https://github.com/scverse/mudata) objects [(Bredikhin et al., 2022)](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02577-8) and extends them to a very simple child class we call **\"MistyData\"**. \n",
    "To make it easier to use, we provide functions to construct \"MistyData\" objects that transform the data into a format that MISTy can use.\n",
    "\n",
    "Briefly, a **\"MistyData\"** object is just a MuData object with **intra** as one of the modalities - this is the view in which the (**target**) variables explained by all other views are stored. \n",
    "MISTy is flexible to any other view that is appended, provided it also contains a spatial neighbors graph.\n",
    "\n",
    "\n",
    "Let's use `genericMistyData` to construct a MuData object with the intra view and the cell type proportions as the first view.\n",
    "Then it additionally build a 'juxta' view for the spots that are neighbors of each other, and a 'para' view for all surrounding spots within a certain radius, or bandwidth.\n",
    "\n",
    "Here, we use the cell type proportions as the intra view."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "misty = genericMistyData(intra=comps, cutoff=0.05, bandwidth=200, n_neighs=6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "misty"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Learn Relationships with MISTy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "misty(model=LinearModel, k_cv=10, seed=1337, verbose = True)\n",
    "\n",
    "#misty(model=RandomForestModel, n_jobs=-1, verbose = True) # you can also use RandomForestModel but it takes longer time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specifically, we will use the `LinearModel` to fit an linear model for each target in the intra view, using the juxta and para views as predictors. It is a bit more simplistic but much faster and more interpretable.\n",
    "You can also use `RandomForestModel` but it takes longer time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MISTy returns two DataFrames:\n",
    "* `target_metrics` - the metrics that describe the target variables from the intra view, including R-squared across different views as well as the estimated contributions to the predictive performance of each view per target.\n",
    "* `interactions` - feature importances per view\n",
    "\n",
    "if `inplace` is true (Default), these are appended to the MuData object.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "check the variance explained when predicting each target variables in the intra view, with other variables (predictors) in the intra view itself\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "misty.uns['target_metrics']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "R² (coefficient of determination) measures how well the model explains the variability of the target variable.\n",
    "\n",
    "intra_R2：R² from the intra-view model\n",
    "\n",
    "multi_R2：R² from the multi-view (full) model\n",
    "\n",
    "gain_R2：multi_R2 - intra_R2, or in other words the performance gain when we additionally consider the other views (in addition to intra). \n",
    "\n",
    "intra, juxta, para: Relative importance (weight) of the intra-view, juxta-view, and para-view."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "li.pl.target_metrics(misty, stat='intra_R2', return_fig=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When we look at the variance explained by the other views in `gain_R2`, we see that they explain less (as expected). There is only a slight gain in Adipocytes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "li.pl.target_metrics(misty, stat='gain_R2', return_fig=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also check the contribution to the predictive performance of each view per target:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "li.pl.contributions(misty, return_fig=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the information above we know which cell types are best explained by our model, and we know which view explains them best."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add pathway activities as extra view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also use estimated pathway activities as extra views to make the data a bit more interpretable.\n",
    "We will use [decoupler-py](https://academic.oup.com/bioinformaticsadvances/article/2/1/vbac016/6544613) with pathways genesets from [PROGENy](https://www.nature.com/articles/s41467-017-02391-6). See [this tutorial](https://decoupler-py.readthedocs.io/en/latest/notebooks/spatial.html) for details."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# obtain genesets\n",
    "progeny = dc.op.progeny(organism='human', top=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use univariate linear model to estimate activity\n",
    "dc.mt.ulm(\n",
    "    adata,\n",
    "    net=progeny,\n",
    "    tmin=5,\n",
    "    verbose=True,\n",
    "    bsize = 256,\n",
    "    raw=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract progeny activities as an AnnData object\n",
    "acts_progeny = li.ut.obsm_to_adata(adata, 'score_ulm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check how the pathway activities look like\n",
    "sc.pl.spatial(acts_progeny, color=['Hypoxia', 'JAK-STAT'], cmap='RdBu_r', size=1.3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this case, we will use cell type compositions per spot as the intra view, and we will use the PROGENy pathway activities to define the juxta and para views:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "misty = genericMistyData(intra=comps, extra=acts_progeny, cutoff=0.05, bandwidth=200, n_neighs=6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "misty"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run MISTy with pathway views"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " we will **bypass** predicting the intraview with features within the intraview features (`bypass_intra`).\n",
    "This will allow us to see how well the other views explain the intraview, excluding the intraview itself."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "misty(model=LinearModel, k_cv=10, seed=1337, bypass_intra=True, verbose = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's check the joined R-squared for views:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "li.pl.target_metrics(misty, stat='gain_R2', return_fig=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the information above we know which variables are best explained by our model, and we know which view explains them best. \n",
    "So, we can now also see what are the specific variables that explain each target best:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this information is stored here:\n",
    "misty.uns['interactions'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "and their contributions per target:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "li.pl.contributions(misty, return_fig=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since this is a linear model, the coefficients would not be directly comparable (as are importances in a Random Forest). Thus, we use the coefficients' t-values, as calculated by Ordinary Least Squares, which are signed and directly comparable.\n",
    "\n",
    "Let's explore the t-values for each target-prediction interaction:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    li.pl.interactions(misty, view='juxta', return_fig=True, figure_size=(7,5)) + \n",
    "    p9.scale_fill_gradient2(low = \"blue\", mid = \"white\", high = \"red\", midpoint = 0)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\">\n",
    "    \n",
    "<h4> Feature importances </h4>\n",
    "\n",
    "By default, we use a random forest, so the feature importances are the mean decrease in Gini impurity of the features. On the other hand, when we use a linear model, the feature importances are the t-values of the model coefficients.\n",
    "\n",
    "\n",
    "</div>  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build Custom Misty Views"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we previously mentioned, one can build any view structure that they deem relevant for their data.\n",
    "So, let's explore how to build custom views.\n",
    "Here, we will just use two distinct prior knowledge sources to check which one achieves better predictive performance."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, let's also estimate Transcription Factor activities with decoupler:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get TF prior knowledge\n",
    "net = dc.op.collectri(organism='human', remove_complexes=True, license='academic', verbose=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Estimate activities\n",
    "dc.mt.ulm(\n",
    "    data=adata,\n",
    "    net=net,\n",
    "    bsize=128,  \n",
    "    tmin = 50,\n",
    "    verbose=True,\n",
    "    raw=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract activities\n",
    "acts_tfs = li.ut.obsm_to_adata(adata, 'score_ulm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# or load acts_tf\n",
    "#acts_tfs = sc.read('acts_tfs.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate spatial neighbors\n",
    "li.ut.spatial_neighbors(acts_tfs, cutoff=0.1, bandwidth=200, set_diag=True) # you can also set set_diag=False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualize the weights for a specific spot:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "li.pl.connectivity(acts_tfs, idx=0, figure_size=(6,5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# transfer spatial information to progeny activities\n",
    "# NOTE: spatial connectivities can differ between views, but in this case we will use the same\n",
    "acts_progeny.obsm['spatial'] = acts_tfs.obsm['spatial']\n",
    "acts_progeny.obsp['spatial_connectivities'] = acts_tfs.obsp['spatial_connectivities']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Build an object with custom views:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "misty = MistyData(data={\"intra\": comps, \"TFs\": acts_tfs, \"Pathways\": acts_progeny})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "misty"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run Misty as before:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "misty(model=LinearModel, k_cv=5, seed=1337, bypass_intra=True, verbose = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that Cardiomyocytes and Fibroblasts are relatively well explained by TFs & Pathways."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "li.pl.target_metrics(misty, stat='gain_R2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also see that both views explain the targets (cell types) to some extent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "li.pl.contributions(misty, return_fig=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot cell type x Trascription factor interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    li.pl.interactions(misty, view='TFs', top_n=20) + \n",
    "    p9.labs(x='Transcription Factor', y='Cell type') +\n",
    "    p9.theme_bw(base_size=14) +\n",
    "    p9.theme(axis_text_x=p9.element_text(rotation=90, size=13)) +\n",
    "    # change to blue-red\n",
    "    p9.scale_fill_gradient2(low='blue', mid='white', high='red')+\n",
    "    p9.theme(figure_size=(8, 5)) \n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Practice\n",
    "\n",
    "Use the use `kuppe_heart19.h5ad` dataset to explore the spatial relationships between cell types and ligand–receptor (LR) interactions, and investigate how different spatial connectivity patterns affect the results.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (02 Liana)",
   "language": "python",
   "name": "02-liana"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
