{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7b3edddb",
   "metadata": {},
   "source": [
    "# NiCo Tutorial"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75135199",
   "metadata": {},
   "source": [
    "## 1. Environment Prep"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b63f057",
   "metadata": {},
   "source": [
    "### 1.1. Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa23a9c4-abc0-499c-953e-4611526762da",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from joblib import Memory\n",
    "import scanpy as sc\n",
    "\n",
    "from nico import Annotations as ann\n",
    "from nico import Interactions as int\n",
    "from nico import Covariations as cov\n",
    "\n",
    "memory = Memory(\"./NiCo/cache\", verbose=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4758aaa",
   "metadata": {},
   "source": [
    "### 1.2. Environment Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16b7feef-39c5-45b8-9b14-834da59b194c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "# in this exercise we need to modify the data so we copy a writeable version into the wd \n",
    "cp -r ../../datasets/Hands_on_3_NiCo/* . "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edc07d38-eecf-4d47-b185-a026b128cfdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_datapath = \"./NiCo/inputQuery/\"\n",
    "ref_datapath = \"./NiCo/inputRef/\"\n",
    "output_nico_dir = \"./NiCo/nico_out/\"\n",
    "lrdb_filename = \"./NiCo/NiCoLRdb.txt\"\n",
    "\n",
    "ref_cluster_tag = \"cluster\"  # scRNAseq cell type slot\n",
    "annotation_slot = \"nico_ct\"  # spatial cell type slot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c23a1d1e-b6b9-4dd7-a508-b0cd1a3000f3",
   "metadata": {},
   "source": [
    "## 2. Preprocessing\n",
    "\n",
    "Results in the following folder structure:\n",
    "\n",
    "```text\n",
    "data/inputQuery\n",
    "â”œâ”€â”€ mouse_intestine_spatial.h5ad -> the original spatial input data\n",
    "â””â”€â”€ sct_spatial.h5ad -> the transformed and subset spatial\n",
    "data/inputRef\n",
    "â”œâ”€â”€ mouse_intestine_sc.h5ad -> the original single cell input data\n",
    "â”œâ”€â”€ Original_counts.h5ad -> the raw counts of the single cell data\n",
    "â””â”€â”€ sct_singleCell.h5ad -> the transformed single cell data\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a77d884",
   "metadata": {},
   "source": [
    "### 2.1. Load and Filter Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb8103f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_query = sc.read_h5ad(query_datapath + \"mouse_intestine_spatial.h5ad\")\n",
    "adata_reference = sc.read_h5ad(ref_datapath + \"mouse_intestine_sc.h5ad\")\n",
    "\n",
    "sc.pp.filter_cells(adata_query, min_counts=5)\n",
    "sc.pp.filter_cells(adata_reference, min_counts=5)\n",
    "\n",
    "sc.pp.filter_genes(adata_query, min_cells=1)\n",
    "sc.pp.filter_genes(adata_reference, min_cells=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2661cff",
   "metadata": {},
   "source": [
    "### 2.2. Save raw reference counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee810543",
   "metadata": {},
   "outputs": [],
   "source": [
    "@memory.cache\n",
    "def save_reference(adata):\n",
    "    Original_counts = adata.copy()\n",
    "    Original_counts.raw = Original_counts.copy()\n",
    "    sc.pp.normalize_total(Original_counts)\n",
    "    sc.pp.log1p(Original_counts)\n",
    "    sc.tl.pca(Original_counts)\n",
    "    sc.pp.neighbors(Original_counts)\n",
    "    #sc.tl.umap(Original_counts)\n",
    "    return Original_counts\n",
    "\n",
    "Original_counts = save_reference(adata_reference)\n",
    "Original_counts.write_h5ad(ref_datapath + \"Original_counts.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cba95956",
   "metadata": {},
   "source": [
    "### 2.3. Perform scTransform on shared gene space"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dff1e98c-3569-41c4-b6d1-e17b7acc6801",
   "metadata": {},
   "outputs": [],
   "source": [
    "index_sp, index_sc = ann.find_index(\n",
    "    adata_query.var_names.to_numpy(), adata_reference.var_names.to_numpy()\n",
    ")\n",
    "\n",
    "ad_seq_common = adata_reference.copy()\n",
    "ad_spatial_common = adata_query[:, index_sp].copy()\n",
    "\n",
    "temp_spatial = ad_spatial_common.copy()\n",
    "temp_seq = ad_seq_common.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f85e1b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_spatial_common = memory.cache(ann.SCTransform)(\n",
    "    ad_spatial_common,\n",
    "    min_cells=1,\n",
    "    gmean_eps=1,\n",
    "    n_genes=500,\n",
    "    n_cells=None,\n",
    "    bin_size=500,\n",
    "    bw_adjust=3,\n",
    "    inplace=False,\n",
    ")\n",
    "\n",
    "ad_seq_common = memory.cache(ann.SCTransform)(\n",
    "    ad_seq_common,\n",
    "    min_cells=1,\n",
    "    gmean_eps=1,\n",
    "    n_genes=500,\n",
    "    n_cells=None,\n",
    "    bin_size=500,\n",
    "    bw_adjust=3,\n",
    "    inplace=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07f26b34-ae9a-4361-9620-7d2df8138885",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_spatial_common.raw = temp_spatial.copy()\n",
    "ad_spatial_common.obsm[\"spatial\"] = temp_spatial.obsm[\"spatial\"]\n",
    "ad_seq_common.raw = temp_seq.copy()\n",
    "\n",
    "ad_seq_common.write_h5ad(\"./NiCo/inputRef/sct_singleCell.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ddb6974",
   "metadata": {},
   "source": [
    "### 2.4. Perform Leiden clustering to guide label transfer\n",
    "\n",
    "The number of clusters should reflect the number of celltypes in the reference we expect to find in the spatial data. It might take some adapation of the Leiden resolution parameter to find a good clustering."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e53030e-0a66-412e-a936-83ebe77c6246",
   "metadata": {},
   "outputs": [],
   "source": [
    "@memory.cache\n",
    "def perform_guiding_clustering(adata, n_pcs: int = 30, leiden_res: float = 0.4):\n",
    "    adata_copy = adata.copy()\n",
    "    sc.pp.pca(adata_copy)\n",
    "    sc.pp.neighbors(adata_copy, n_pcs=n_pcs)\n",
    "    sc.tl.umap(adata_copy)\n",
    "    key_added = f\"leiden{leiden_res}\"\n",
    "    sc.tl.leiden(adata_copy, resolution=leiden_res, key_added=f\"leiden{leiden_res}\")\n",
    "    return adata_copy\n",
    "    \n",
    "leiden_res = 0.4\n",
    "ad_spatial_common = perform_guiding_clustering(ad_spatial_common, leiden_res=leiden_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffac897c",
   "metadata": {},
   "outputs": [],
   "source": [
    "key_added = f\"leiden{leiden_res}\"\n",
    "sc.pl.umap(\n",
    "    ad_spatial_common,\n",
    "    color=[key_added],\n",
    "    title=[\"Spatial umap\"],\n",
    "    wspace=0.4,\n",
    "    show=True,\n",
    "    save=\"_spatial_umap.png\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d4dfa46",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_spatial_common.write_h5ad(query_datapath + \"sct_spatial.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69e58d4a",
   "metadata": {},
   "source": [
    "## 3. Label Transfer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fc4a44a",
   "metadata": {},
   "source": [
    "### 3.1. Computation\n",
    "\n",
    "1. Principal components are computed based on the (transformed) single cell reference.\n",
    "2. The (transformed) cells from the spatial dataset are projected into this latent space and based on an Euclidian distance metric cells are identified that 'anchor' the spatial data with the single cell reference.\n",
    "3. Labels are propagated iteratively from the anchor cells to neighbouring cells in the latent space, with guidance from the Leiden clustering.\n",
    "4. Annotation saved"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2035a810",
   "metadata": {},
   "outputs": [],
   "source": [
    "anchors_and_neighbors_info = memory.cache(ann.find_anchor_cells_between_ref_and_query)(\n",
    "    refpath=ref_datapath,\n",
    "    quepath=query_datapath,\n",
    "    output_nico_dir=output_nico_dir,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e52647fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_info = memory.cache(ann.nico_based_annotation)(\n",
    "    anchors_and_neighbors_info,\n",
    "    guiding_spatial_cluster_resolution_tag=key_added,\n",
    "    ref_cluster_tag=ref_cluster_tag,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ce97f4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ann.save_annotations_in_spatial_object(output_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d44f12ca-9613-4d8e-97a8-e9513623baa1",
   "metadata": {},
   "source": [
    "### 3.2. Visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a3269a7",
   "metadata": {},
   "source": [
    "#### 3.2.1. Full annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00dc432b-455c-4896-84f2-a43fe4770c7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ann.visualize_umap_and_cell_coordinates_with_all_celltypes(\n",
    "    output_nico_dir=output_nico_dir,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8ac735d",
   "metadata": {},
   "source": [
    "#### 3.2.2. Filtered for cells of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50a8aa1e-a7a6-4c6f-a73f-7b741e4bdef7",
   "metadata": {},
   "outputs": [],
   "source": [
    "choose_celltypes = [[\"Stem/TA\", \"Paneth\"], [\"Paneth\", \"Goblet\"]]\n",
    "\n",
    "ann.visualize_umap_and_cell_coordinates_with_selected_celltypes(\n",
    "    choose_celltypes=choose_celltypes,\n",
    "    output_nico_dir=output_nico_dir,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c3d97ae-7a1f-4661-97f3-f950f16beb0e",
   "metadata": {},
   "source": [
    "## 4. Niche Interactions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71b8a472",
   "metadata": {},
   "source": [
    "### 4.1. Computation\n",
    "\n",
    "1. Graph is computed based on the spatial coordinates and a given radius\n",
    "2. Local neighbourhood composition for each central cell is computed\n",
    "3. Interaction coefficients are estimated via logarithmic regression\n",
    "\n",
    "In this example we exclude celltypes with a very low abundance from the analysis.\n",
    "\n",
    "**Note on Radius Definition**\n",
    "\n",
    "The radius parameter (R) in NiCo controls how spatial neighbourhoods are defined around each central cell.\n",
    "- When R = 0, NiCo considers only the immediate neighbours of each cell as determined by Delaunay triangulation. This setting focuses on direct cellâ€“cell communication within tissues.\n",
    "- When R > 0, NiCo extends the neighbourhood to include all cells within a specified distance from the central cell. This can capture paracrine or diffusive signalling, where molecules act over finite spatial ranges.\n",
    "\n",
    "In practice, we recommend starting with R = 0, since:\n",
    "- Immediate neighbours usually capture the strongest covariation and interaction signals.\n",
    "- Increasing R smooths or averages signals across a larger set of cells, which can dilute the signal of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d80f4fd7-72d1-462e-9307-d44b87d90b12",
   "metadata": {},
   "outputs": [],
   "source": [
    "inputRadius = 0\n",
    "\n",
    "niche_pred_output = memory.cache(int.spatial_neighborhood_analysis)(\n",
    "    Radius=inputRadius,\n",
    "    output_nico_dir=output_nico_dir,\n",
    "    removed_CTs_before_finding_CT_CT_interactions=[\n",
    "        \"Basophils\",\n",
    "        \"Cycling/GC B cell\",\n",
    "        \"pDC\",\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea35fb96-2cf2-433a-b878-e1f513a392ad",
   "metadata": {},
   "source": [
    "### 4.2. Visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea432c34",
   "metadata": {},
   "source": [
    "#### 4.2.1. Plot interaction graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41331154-36c1-47dc-a164-a6951af1996f",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_niche_interaction_cutoff = 0.1\n",
    "\n",
    "int.plot_niche_interactions_with_edge_weight(\n",
    "    niche_pred_output,\n",
    "    niche_cutoff=celltype_niche_interaction_cutoff,\n",
    "    node_size=1000,\n",
    "    node_font_size=10,\n",
    "    font_weight=\"bold\",\n",
    "    edge_font_size=5,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34ecd458-a208-4ccc-9f2b-701d37d0724a",
   "metadata": {},
   "source": [
    "#### 4.2.2. Individual cell type niche plot\n",
    "To plot regression coefficients of niche cell types for given central cell types, ordered by magnitude, <br>\n",
    "add cell type names for the desired central cell types to the list argument choose_celltypes (e.g. Stem/TA <br>\n",
    "and Paneth cells)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d8cbbe2-f285-4afb-af37-ea141d6ebaca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Blue dotted line in following plot indicates the celltype_niche_interaction_cutoff\n",
    "\n",
    "int.find_interacting_cell_types(\n",
    "    niche_pred_output,\n",
    "    choose_celltypes=[\"Stem/TA\", \"Paneth\"],\n",
    "    celltype_niche_interaction_cutoff=celltype_niche_interaction_cutoff,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5648e71c-fa25-41e0-9d9b-a22b70ee1eb9",
   "metadata": {},
   "source": [
    "## 5. Latent Feature Covariation Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bab37d0",
   "metadata": {},
   "source": [
    "### 5.1. Computation\n",
    "\n",
    "1. Latent spaces (default dimension = 3) are computed for each celltype via NMF, either with data integration or on single cell data only with projection of spatial data\n",
    "2. Latent features are aggregated on the spatial neighbourhood graph and used to predict a neighbourhoods central cells latent features via Ridge regression.\n",
    "The aggregation is optionally filtered as per the coefficients computed in the Interaction module, so that only positively interacting cell types are used.\n",
    "3. Explainability of latent feature loadings is extended via correlation with the full gene set present in the reference data.\n",
    "4. Ligand-Receptor analysis between two latent feature loadings is performed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1287764-9ad1-4779-930c-68dc75524335",
   "metadata": {},
   "source": [
    "<div style=\"\n",
    "    background-color: #ffded1;\n",
    "    border-left: 4px solid #ff0000;\n",
    "    padding: 10px 15px;\n",
    "    border-radius: 5px;\n",
    "    margin: 10px 0;\n",
    "\">\n",
    "<b>ðŸ’¡ Warning:</b> This computation takes a while so we are caching.\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d97f5d58-47fb-49eb-b501-1cb02c2ac457",
   "metadata": {},
   "outputs": [],
   "source": [
    "cov_out = memory.cache(cov.gene_covariation_analysis)(\n",
    "    iNMFmode=True,  # Adjust in case of error prone segmentation\n",
    "    Radius=inputRadius,\n",
    "    refpath=ref_datapath,\n",
    "    quepath=query_datapath,\n",
    "    output_niche_prediction_dir=output_nico_dir,\n",
    "    ref_cluster_tag=ref_cluster_tag,\n",
    "    LRdbFilename=lrdb_filename,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05c8f4b1",
   "metadata": {},
   "source": [
    "### 5.2. Visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64102338-e6c5-4bcf-b56f-8b02a10c1061",
   "metadata": {},
   "source": [
    "#### 5.2.1. Visualize the cosine similarity and Spearman correlation between genes and latent factors\n",
    "\n",
    "The following function generates output for the top 30 genes based on cosine similarity (left) or Spearman correlation (right) with latent factors.\n",
    "\n",
    "Select cell types by adding IDs to the list argument choose_celltypes, or leave empty for generating output for all cell types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d7164fb-07ab-48c2-86a2-5df22a708f32",
   "metadata": {},
   "outputs": [],
   "source": [
    "cov.plot_cosine_and_spearman_correlation_to_factors(\n",
    "    cov_out,\n",
    "    choose_celltypes=[\"Paneth\"],\n",
    "    figsize=(15, 10),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "686754ba-a456-4ad8-be89-9c6281b5459a",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataFrame = cov.extract_and_plot_top_genes_from_chosen_factor_in_celltype(\n",
    "    cov_out,\n",
    "    choose_celltype=\"Stem/TA\",\n",
    "    choose_factor_id=1,\n",
    "    top_NOG=20,\n",
    "    correlation_with_spearman=True,\n",
    "    positively_correlated=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cc02e10-014f-4f62-acff-2c8e7b89fe66",
   "metadata": {},
   "source": [
    "#### 5.2.2. Cell type covariation visualization\n",
    "\n",
    "Plot linear regression coefficients between factors of the central cell type (y-axis, defined by list argument choose_celltypes) and factors of niche cell types (x-axis).\n",
    "\n",
    "Circle size scales with -log10(p-value) (indicated as number on top of each circle). To generate plots for all cell types, leave list argument choose_celltypes empty.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6766748-8730-4f25-88fb-98ebff33cdc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "choose_celltypes = [\"Stem/TA\"]\n",
    "cov.plot_significant_regression_covariations_as_circleplot(\n",
    "    cov_out,\n",
    "    choose_celltypes=choose_celltypes,\n",
    "    mention_pvalue=True,\n",
    "    figsize=(6, 1.25),\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e27a954",
   "metadata": {},
   "source": [
    "#### 5.2.3. Analysis of ligand-receptor interactions between covarying niche cell types\n",
    "\n",
    "Perform ligand-receptors analysis. In this example, output is generated for the ligand-receptor pairs associated with the interacting factor 1 of Stem/TA cells and factor 1 of Paneth cells.\n",
    "\n",
    "`choose_interacting_celltype_pair = ['Stem/TA', 'Paneth']`\n",
    "\n",
    "`choose_factors_id = [1, 1]` entries correspond to cell types in `choose_interacting_celltype_pair`, i.e., first factor ID corresponds to `Stem/TA` and second factor ID corresponds to `Paneth`.\n",
    "\n",
    "By default, the analysis is saved in 3 separate figures (bidirectional, CC to NC and NC to CC). CC: central cell, NC: niche cell.\n",
    "\n",
    "Our analysis accounts for bidirectional cellular crosstalk interactions of ligands and receptors in cell types A and B. The ligand can be expressed on cell type A and signal to the receptor detected on cell type B, or vice versa.\n",
    "\n",
    "By changing the cutoff for minimum factor correlation of ligand/receptor genes (`LR_plot_NMF_Fa_thres = 0.2`) or the cutoff for the minimum fraction of cells expressing the ligand/receptor genes (`LR_plot_Exp_thres = 0.2`), the stringency of the output filtering can be controlled."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1e5c79b-9391-4213-974c-0a951fbbb427",
   "metadata": {},
   "outputs": [],
   "source": [
    "cov.find_LR_interactions_in_interacting_cell_types(\n",
    "    cov_out,\n",
    "    choose_interacting_celltype_pair=[\"Stem/TA\", \"Paneth\"],\n",
    "    choose_factors_id=[2, 1],\n",
    "    pvalueCutoff=0.05,\n",
    "    LR_plot_NMF_Fa_thres=0.3,\n",
    "    LR_plot_Exp_thres=0.2,\n",
    "    figsize=(12, 10),\n",
    "    showit=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3937692d",
   "metadata": {},
   "source": [
    "#### 5.2.4. Functional enrichment analysis for genes associated with latent factors\n",
    "\n",
    "This step performs **pathway enrichment analysis** for genes most strongly associated with a selected latent factor.\n",
    "\n",
    "In this example, enrichment is carried out for the **top 50** genes (`NOG_pathway=50`) that are **positively correlated** (`positively_correlated=True`) with **factor 2** (`choose_factors_id=[2]`) of **Stem/TA cells** (`choose_celltypes=['Stem/TA']`). The analysis tests for overrepresentation of **Gene Ontology (GO) Biological Processes** using the database specified by `database=['GO_Biological_Process_2021']`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ae33505-0b09-45f7-afc0-0e2373145bd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "cov.pathway_analysis(\n",
    "    cov_out,\n",
    "    choose_celltypes=[\"Stem/TA\"],\n",
    "    NOG_pathway=50,\n",
    "    choose_factors_id=[2],\n",
    "    positively_correlated=True,\n",
    "    database=[\"GO_Biological_Process_2021\"],  # database=['BioPlanet_2019'],\n",
    "    rps_rpl_mt_genes_included=False,\n",
    "    display_plot_as=\"barplot\",\n",
    "    correlation_with_spearman=True,\n",
    "    object_for_color=\"Adjusted P-value\",\n",
    "    object_for_xaxis=\"Odds Ratio\",\n",
    "    savefigure=False,\n",
    "    input_colormap=\"hot_r\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "075ca75e-3bcd-4097-8347-d2f5e8054c9d",
   "metadata": {},
   "source": [
    "#### 5.2.5. Visualization of top genes across cell types and factors as dotplot\n",
    "\n",
    "Show the top 20 positively and negatively correlated genes (top_NOG=20) for all latent factors and the average expression of these genes on a log scale in a single plot. In this example, plots are generated for Paneth and Stem/TA cells.\n",
    "\n",
    "If ```choose_celltypes=[]```, the plot will be generated for all cell\n",
    "types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b79bc205-f739-467c-8108-9a54a63211b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "cov.plot_top_genes_for_a_given_celltype_from_all_factors(\n",
    "    cov_out,\n",
    "    choose_celltypes=[\"Paneth\", \"Stem/TA\"],\n",
    "    top_NOG=20,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (03 Nico)",
   "language": "python",
   "name": "03-nico"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
