{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "pj5css8NI-Cw"
   },
   "source": [
    "# Introduction to Spatial Omics with 10X Xenium and Visium\n",
    "\n",
    "\n",
    "## The breast cancer dataset\n",
    "\n",
    "## Setup\n",
    "\n",
    "\n",
    "### Download data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "B5zAOlJWFvqF"
   },
   "source": [
    "### Required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "import pathlib\n",
    "import numpy\n",
    "import scanpy\n",
    "import matplotlib\n",
    "import seaborn\n",
    "import scipy\n",
    "import sklearn # scikit-learn is imported as sklearn\n",
    "import igraph as igraph_lib # python-igraph is often imported as igraph\n",
    "import leidenalg\n",
    "import squidpy\n",
    "import zarr\n",
    "import numcodecs\n",
    "import anndata \n",
    "import h5py\n",
    "\n",
    "\n",
    "print(\"--- Installed Package Versions ---\")\n",
    "print(f\"pandas:          {pandas.__version__}\")\n",
    "print(f\"pathlib:         Standard Library (no version attribute)\")\n",
    "print(f\"numpy:           {numpy.__version__}\")\n",
    "print(f\"scanpy:          {scanpy.__version__}\")\n",
    "print(f\"matplotlib:      {matplotlib.__version__}\")\n",
    "print(f\"seaborn:         {seaborn.__version__}\")\n",
    "print(f\"scipy:           {scipy.__version__}\")\n",
    "print(f\"scikit-learn:    {sklearn.__version__}\")\n",
    "print(f\"python-igraph:   {igraph_lib.__version__}\")\n",
    "print(f\"leidenalg:       {leidenalg.__version__}\")\n",
    "print(f\"squidpy:         {squidpy.__version__}\")\n",
    "print(f\"zarr:            {zarr.__version__}\")\n",
    "print(f\"numcodecs:       {numcodecs.__version__}\")\n",
    "print(f\"anndata:         {anndata.__version__}\")\n",
    "print(f\"h5py:            {h5py.__version__}\")\n",
    "print(\"----------------------------------\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "4imeiSQXvuR5"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import os\n",
    "from collections import namedtuple, Counter\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "# import squidpy as sq\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy.spatial import KDTree\n",
    "from matplotlib import rcParams\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from typing import Dict, Tuple\n",
    "FIGSIZE=(5,5)\n",
    "rcParams['figure.figsize']=FIGSIZE\n",
    "from typing import Tuple, Union, List, Optional, Dict\n",
    "\n",
    "import anndata as ad\n",
    "import squidpy as sq\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "from sklearn.neighbors import KDTree\n",
    "import numpy as np\n",
    "from scipy.sparse import lil_matrix\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "datadir = \"../../datasets/Hands_on_1_Spatial_Data/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VRg_B1OWagrF",
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Required functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 211
    },
    "id": "dQGcav_IagrG",
    "outputId": "2540b0ad-8c7e-43e2-c83c-31637506682b"
   },
   "outputs": [],
   "source": [
    "from sklearn.metrics import calinski_harabasz_score, silhouette_score, f1_score\n",
    "def compute_clustering_scores(\n",
    "    adata: sc.AnnData, cell_type_column: str = \"celltype_major\", use_pca: bool = True\n",
    ") -> Tuple[float, float]:\n",
    "    \"\"\"Compute the Calinski-Harabasz and Silhouette scores for an AnnData object.\"\"\"\n",
    "    if cell_type_column not in adata.obs:\n",
    "        raise ValueError(f\"Column '{cell_type_column}' must be present in adata.obs.\")\n",
    "\n",
    "    # Randomly sample a subset for efficiency (e.g., 10,000 cells)\n",
    "    n_cells = min(adata.n_obs, 10000)  # Limit to 10k cells\n",
    "    cell_indices = np.random.choice(adata.n_obs, n_cells, replace=False)\n",
    "\n",
    "    # Select features\n",
    "    features = adata.obsm[\"X_pca\"] if use_pca else adata.X\n",
    "    features = features[cell_indices, :]\n",
    "    labels = adata[cell_indices, :].obs[cell_type_column].values\n",
    "\n",
    "    # Compute scores\n",
    "    ch_score = calinski_harabasz_score(features, labels)\n",
    "    sh_score = silhouette_score(features, labels)\n",
    "    return ch_score, sh_score\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def plot_embedding(\n",
    "    adata: sc.AnnData,\n",
    "    features: Union[str, List[str]],\n",
    "    basis: str = 'X_umap',\n",
    "    palette: str = \"tab20\",\n",
    "    cmap: str = \"plasma\",\n",
    "    titles: Union[str, List[str]] = None,\n",
    "    ncols: int = 5,\n",
    "    dim: int = 5,\n",
    "    layer: str = \"imputed\",\n",
    "    dim_label = \"UMAP\",\n",
    "    ax = None,\n",
    "    fs: int = 12,\n",
    "    lw: float = 1.5,\n",
    "    arrow_len: float = 0.2,\n",
    "    draw_arrows=False,\n",
    "    rasterized=False,\n",
    "    legend=False,\n",
    "    **kwargs,\n",
    "):\n",
    "    iterify = lambda x: x if isinstance(x, Iterable) and not isinstance(x, str) else [x]\n",
    "    features = iterify(features)\n",
    "    titles = iterify(titles)\n",
    "    if not ax:\n",
    "        nrows = math.ceil(len(features))\n",
    "        fig, axes = plt.subplots(\n",
    "            nrows, ncols,\n",
    "            figsize=(dim*ncols,dim*nrows),\n",
    "        )\n",
    "        fig.tight_layout(pad=dim*0.75)\n",
    "        axes = axes.flat if isinstance(axes, Iterable) else [axes]\n",
    "    else:\n",
    "        assert (len(features)==1) and (len(titles)==1)\n",
    "        fig = ax.get_figure()\n",
    "        axes = [ax]\n",
    "    if legend:\n",
    "        legend_loc='right margin'\n",
    "    else:\n",
    "        legend_loc=None\n",
    "    for ax, feature, title in itertools.zip_longest(axes, features, titles):\n",
    "        if not title: title = feature\n",
    "        if feature:\n",
    "            sc.pl.embedding(\n",
    "                adata,\n",
    "                basis=basis,\n",
    "                color=feature,\n",
    "                ax=ax,\n",
    "                show=False,\n",
    "                palette=palette, cmap=cmap,\n",
    "                layer=layer,\n",
    "                legend_loc=legend_loc,\n",
    "                **kwargs,\n",
    "            )\n",
    "            if rasterized:\n",
    "                ax.get_children()[0].set_rasterized(True)\n",
    "            format_ax(\n",
    "                fig, ax, style=\"umap\",\n",
    "                title=title, dim_label=dim_label, fs=fs,\n",
    "                arrow_len=arrow_len, lw=lw, draw_arrows=draw_arrows,\n",
    "            )\n",
    "        else:\n",
    "            ax.set_visible(False)\n",
    "    return fig\n",
    "\n",
    "\n",
    "\n",
    "def find_markers(\n",
    "    adata: ad.AnnData,\n",
    "    cell_type_column: str,\n",
    "    pos_percentile: float = 5,\n",
    "    neg_percentile: float = 10,\n",
    "    percentage: float = 50,\n",
    ") -> Dict[str, Dict[str, List[str]]]:\n",
    "    \"\"\"Identify positive and negative markers for each cell type based on gene expression and filter by expression percentage.\n",
    "\n",
    "    Args:\n",
    "    - adata: AnnData\n",
    "        Annotated data object containing gene expression data.\n",
    "    - cell_type_column: str\n",
    "        Column name in `adata.obs` that specifies cell types.\n",
    "    - pos_percentile: float, default=5\n",
    "        Percentile threshold to determine top x% expressed genes.\n",
    "    - neg_percentile: float, default=10\n",
    "        Percentile threshold to determine top x% lowly expressed genes.\n",
    "    - percentage: float, default=50\n",
    "        Minimum percentage of cells expressing the marker within a cell type for it to be considered.\n",
    "\n",
    "    Returns:\n",
    "    - markers: dict\n",
    "        Dictionary where keys are cell types and values are dictionaries containing:\n",
    "            'positive': list of top x% highly expressed genes\n",
    "            'negative': list of top x% lowly expressed genes.\n",
    "    \"\"\"\n",
    "    markers = {}\n",
    "    sc.tl.rank_genes_groups(adata, groupby=cell_type_column)\n",
    "    genes = adata.var_names\n",
    "    for cell_type in adata.obs[cell_type_column].unique():\n",
    "        subset = adata[adata.obs[cell_type_column] == cell_type]\n",
    "        mean_expression = np.asarray(subset.X.mean(axis=0)).flatten()\n",
    "        cutoff_high = np.percentile(mean_expression, 100 - pos_percentile)\n",
    "        cutoff_low = np.percentile(mean_expression, neg_percentile)\n",
    "        pos_indices = np.where(mean_expression >= cutoff_high)[0]\n",
    "        neg_indices = np.where(mean_expression <= cutoff_low)[0]\n",
    "        expr_frac = np.asarray((subset.X[:, pos_indices] > 0).mean(axis=0)).flatten()\n",
    "        valid_pos_indices = pos_indices[expr_frac >= (percentage / 100)]\n",
    "        positive_markers = genes[valid_pos_indices]\n",
    "        negative_markers = genes[neg_indices]\n",
    "        markers[cell_type] = {\"positive\": list(positive_markers), \"negative\": list(negative_markers)}\n",
    "    return markers\n",
    "\n",
    "\n",
    "def find_mutually_exclusive_genes(\n",
    "    adata: ad.AnnData, markers: Dict[str, Dict[str, List[str]]], cell_type_column: str\n",
    ") -> List[Tuple[str, str]]:\n",
    "    \"\"\"Identify mutually exclusive genes based on expression criteria.\n",
    "\n",
    "    Args:\n",
    "    - adata: AnnData\n",
    "        Annotated data object containing gene expression data.\n",
    "    - markers: dict\n",
    "        Dictionary where keys are cell types and values are dictionaries containing:\n",
    "            'positive': list of top x% highly expressed genes\n",
    "            'negative': list of top x% lowly expressed genes.\n",
    "    - cell_type_column: str\n",
    "        Column name in `adata.obs` that specifies cell types.\n",
    "\n",
    "    Returns:\n",
    "    - exclusive_pairs: list\n",
    "        List of mutually exclusive gene pairs.\n",
    "    \"\"\"\n",
    "    exclusive_genes = {}\n",
    "    all_exclusive = []\n",
    "    gene_expression = adata.to_df()\n",
    "    for cell_type, marker_sets in markers.items():\n",
    "        positive_markers = marker_sets[\"positive\"]\n",
    "        exclusive_genes[cell_type] = []\n",
    "        for gene in positive_markers:\n",
    "            gene_expr = adata[:, gene].X\n",
    "            cell_type_mask = adata.obs[cell_type_column] == cell_type\n",
    "            non_cell_type_mask = ~cell_type_mask\n",
    "            if (gene_expr[cell_type_mask] > 0).mean() > 0.2 and (gene_expr[non_cell_type_mask] > 0).mean() < 0.05:\n",
    "                exclusive_genes[cell_type].append(gene)\n",
    "                all_exclusive.append(gene)\n",
    "    unique_genes = list({gene for i in exclusive_genes.keys() for gene in exclusive_genes[i] if gene in all_exclusive})\n",
    "    filtered_exclusive_genes = {\n",
    "        i: [gene for gene in exclusive_genes[i] if gene in unique_genes] for i in exclusive_genes.keys()\n",
    "    }\n",
    "    mutually_exclusive_gene_pairs = [\n",
    "        (gene1, gene2)\n",
    "        for key1, key2 in combinations(filtered_exclusive_genes.keys(), 2)\n",
    "        for gene1 in filtered_exclusive_genes[key1]\n",
    "        for gene2 in filtered_exclusive_genes[key2]\n",
    "    ]\n",
    "    return mutually_exclusive_gene_pairs\n",
    "\n",
    "\n",
    "def compute_MECR(adata: ad.AnnData, gene_pairs: List[Tuple[str, str]]) -> Dict[Tuple[str, str], float]:\n",
    "    \"\"\"Compute the Mutually Exclusive Co-expression Rate (MECR) for each gene pair in an AnnData object.\n",
    "\n",
    "    Args:\n",
    "    - adata: AnnData\n",
    "        Annotated data object containing gene expression data.\n",
    "    - gene_pairs: List[Tuple[str, str]]\n",
    "        List of tuples representing gene pairs to evaluate.\n",
    "\n",
    "    Returns:\n",
    "    - mecr_dict: Dict[Tuple[str, str], float]\n",
    "        Dictionary where keys are gene pairs (tuples) and values are MECR values.\n",
    "    \"\"\"\n",
    "    mecr_dict = {}\n",
    "    gene_expression = adata.to_df()\n",
    "    for gene1, gene2 in gene_pairs:\n",
    "        expr_gene1 = gene_expression[gene1] > 0\n",
    "        expr_gene2 = gene_expression[gene2] > 0\n",
    "        both_expressed = (expr_gene1 & expr_gene2).mean()\n",
    "        at_least_one_expressed = (expr_gene1 | expr_gene2).mean()\n",
    "        mecr = both_expressed / at_least_one_expressed if at_least_one_expressed > 0 else 0\n",
    "        mecr_dict[(gene1, gene2)] = mecr\n",
    "    return mecr_dict\n",
    "\n",
    "\n",
    "def compute_quantized_mecr_area(\n",
    "    adata: sc.AnnData, gene_pairs: List[Tuple[str, str]], quantiles: int = 10\n",
    ") -> pd.DataFrame:\n",
    "    \"\"\"Compute the average MECR, variance of MECR, and average cell area for quantiles of cell areas.\n",
    "\n",
    "    Args:\n",
    "    - adata: AnnData\n",
    "        Annotated data object containing gene expression data.\n",
    "    - gene_pairs: List[Tuple[str, str]]\n",
    "        List of tuples representing gene pairs to evaluate.\n",
    "    - quantiles: int, default=10\n",
    "        Number of quantiles to divide the data into.\n",
    "\n",
    "    Returns:\n",
    "    - quantized_data: pd.DataFrame\n",
    "        DataFrame containing quantile information, average MECR, variance of MECR, average area, and number of cells.\n",
    "    \"\"\"\n",
    "    adata.obs[\"quantile\"] = pd.qcut(adata.obs[\"cell_area\"], quantiles, labels=False)\n",
    "    quantized_data = []\n",
    "    for quantile in range(quantiles):\n",
    "        cells_in_quantile = adata.obs[\"quantile\"] == quantile\n",
    "        mecr = compute_MECR(adata[cells_in_quantile, :], gene_pairs)\n",
    "        average_mecr = np.mean([i for i in mecr.values()])\n",
    "        variance_mecr = np.var([i for i in mecr.values()])\n",
    "        average_area = adata.obs.loc[cells_in_quantile, \"cell_area\"].mean()\n",
    "        quantized_data.append(\n",
    "            {\n",
    "                \"quantile\": quantile / quantiles,\n",
    "                \"average_mecr\": average_mecr,\n",
    "                \"variance_mecr\": variance_mecr,\n",
    "                \"average_area\": average_area,\n",
    "                \"num_cells\": cells_in_quantile.sum(),\n",
    "            }\n",
    "        )\n",
    "    return pd.DataFrame(quantized_data)\n",
    "\n",
    "\n",
    "def compute_quantized_mecr_counts(\n",
    "    adata: sc.AnnData, gene_pairs: List[Tuple[str, str]], quantiles: int = 10\n",
    ") -> pd.DataFrame:\n",
    "    \"\"\"Compute the average MECR, variance of MECR, and average transcript counts for quantiles of transcript counts.\n",
    "\n",
    "    Args:\n",
    "    - adata: AnnData\n",
    "        Annotated data object containing gene expression data.\n",
    "    - gene_pairs: List[Tuple[str, str]]\n",
    "        List of tuples representing gene pairs to evaluate.\n",
    "    - quantiles: int, default=10\n",
    "        Number of quantiles to divide the data into.\n",
    "\n",
    "    Returns:\n",
    "    - quantized_data: pd.DataFrame\n",
    "        DataFrame containing quantile information, average MECR, variance of MECR, average counts, and number of cells.\n",
    "    \"\"\"\n",
    "    adata.obs[\"quantile\"] = pd.qcut(adata.obs[\"transcripts\"], quantiles, labels=False)\n",
    "    quantized_data = []\n",
    "    for quantile in range(quantiles):\n",
    "        print(quantile)\n",
    "        cells_in_quantile = adata.obs[\"quantile\"] == quantile\n",
    "        mecr = compute_MECR(adata[cells_in_quantile, :], gene_pairs)\n",
    "        average_mecr = np.mean([i for i in mecr.values()])\n",
    "        variance_mecr = np.var([i for i in mecr.values()])\n",
    "        average_counts = adata.obs.loc[cells_in_quantile, \"transcripts\"].mean()\n",
    "        quantized_data.append(\n",
    "            {\n",
    "                \"quantile\": quantile / quantiles,\n",
    "                \"average_mecr\": average_mecr,\n",
    "                \"variance_mecr\": variance_mecr,\n",
    "                \"average_counts\": average_counts,\n",
    "                \"num_cells\": cells_in_quantile.sum(),\n",
    "            }\n",
    "        )\n",
    "    return pd.DataFrame(quantized_data)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def calculate_sensitivity(\n",
    "    adata: ad.AnnData, purified_markers: Dict[str, Dict[str, List[str]]], max_cells_per_type: int = 1000\n",
    ") -> Dict[str, List[float]]:\n",
    "    \"\"\"Calculate the sensitivity of purified markers for each cell type based on non-negative markers.\n",
    "    Args:\n",
    "    - adata: AnnData\n",
    "        Annotated data object containing gene expression data.\n",
    "    - purified_markers: dict\n",
    "        Dictionary where keys are cell types, and values are dictionaries with keys\n",
    "        \"positive\" (genes to be expressed) and \"negative\" (genes not to be expressed).\n",
    "    - max_cells_per_type: int, default=1000\n",
    "        Maximum number of cells to consider per cell type.\n",
    "\n",
    "    Returns:\n",
    "    - sensitivity_results: dict\n",
    "        Dictionary with cell types as keys and lists of sensitivity values for each cell.\n",
    "    \"\"\"\n",
    "    sensitivity_results = {cell_type: [] for cell_type in purified_markers.keys()}\n",
    "\n",
    "    for cell_type, markers in purified_markers.items():\n",
    "        positive_markers = markers[\"positive\"]\n",
    "\n",
    "        subset = adata[adata.obs[\"celltype_major\"] == cell_type]\n",
    "\n",
    "        if subset.n_obs > max_cells_per_type:\n",
    "            cell_indices = np.random.choice(subset.n_obs, max_cells_per_type, replace=False)\n",
    "            subset = subset[cell_indices]\n",
    "\n",
    "        for cell_counts in subset.X:\n",
    "            # Total counts for all positive and positive markers\n",
    "            positive_indices = subset.var_names.get_indexer(positive_markers)\n",
    "            total_counts = cell_counts.sum()\n",
    "            positive_counts = cell_counts[positive_indices].sum() if positive_markers else 0\n",
    "            # Sensitivity: Proportion of counts from non-positive markers\n",
    "            sensitivity = positive_counts / total_counts if total_counts > 0 else 0\n",
    "            sensitivity_results[cell_type].append(sensitivity)\n",
    "\n",
    "    return sensitivity_results\n",
    "\n",
    "\n",
    "\n",
    "def plot_quantized_mecr_counts(\n",
    "    quantized_mecr_counts: Dict[str, pd.DataFrame], palette: Dict[str, str]\n",
    ") -> None:\n",
    "    \"\"\"Plot the quantized MECR values against transcript counts for each segmentation method,\n",
    "    with point size proportional to the variance of MECR.\n",
    "\n",
    "    Args:\n",
    "        quantized_mecr_counts (Dict[str, pd.DataFrame]): Dictionary of quantized MECR count data for each segmentation method.\n",
    "        palette (Dict[str, str]): Dictionary mapping segmentation method names to color codes.\n",
    "    \"\"\"\n",
    "    # Apply Arial font\n",
    "    plt.rcParams['font.family'] = 'Arial'\n",
    "\n",
    "    # Create the plot\n",
    "    plt.figure(figsize=(3, 3))  # Slim plot\n",
    "    for method, df in quantized_mecr_counts.items():\n",
    "        # Plot lines\n",
    "        plt.plot(\n",
    "            df[\"average_counts\"],\n",
    "            df[\"average_mecr\"],\n",
    "            marker=\"o\",\n",
    "            linestyle=\"-\",\n",
    "            color=palette.get(method, \"#333333\"),\n",
    "            label=method,\n",
    "            markersize=0,  # No markers, only lines\n",
    "        )\n",
    "        # Scatter points with size proportional to variance\n",
    "        plt.scatter(\n",
    "            df[\"average_counts\"],\n",
    "            df[\"average_mecr\"],\n",
    "            s=df[\"variance_mecr\"] * 8 * 1e4,  # Size of points based on variance\n",
    "            color=palette.get(method, \"#333333\"),\n",
    "            alpha=1,  # Transparency for overlapping points\n",
    "            edgecolor=\"w\",  # White edge for better visibility\n",
    "            linewidth=0.5,  # Thin edge lines\n",
    "        )\n",
    "\n",
    "    # Styling\n",
    "    plt.title(\"\", fontsize=12, pad=10)\n",
    "    plt.xlabel(\"Average Transcript Counts\", fontsize=10)\n",
    "    plt.ylabel(\"Quantized MECR by Transcript Count\", fontsize=10)\n",
    "\n",
    "    # Set y-axis ticks to 2-3 values with .2f format\n",
    "    ymin, ymax = plt.ylim()\n",
    "    y_ticks = [ymin, (ymin + ymax) / 3, (ymin + ymax) / 1.5, ymax]\n",
    "    plt.yticks(y_ticks, [f\"{tick:.2f}\" for tick in y_ticks], fontsize=8)\n",
    "\n",
    "    # Set x-axis ticks for better readability\n",
    "    plt.xticks(fontsize=8)\n",
    "\n",
    "    # Remove horizontal grid and minimize clutter\n",
    "    ax = plt.gca()\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['left'].set_linewidth(1.5)  # Emphasize y-axis\n",
    "    ax.spines['bottom'].set_linewidth(1.5)  # Emphasize x-axis\n",
    "\n",
    "    # plt.grid(False)\n",
    "    # sns.despine(left=True, bottom=True)\n",
    "\n",
    "    # Adjust legend placement outside the plot\n",
    "    plt.legend(\n",
    "        title=\"\",\n",
    "        bbox_to_anchor=(1.05, 1),\n",
    "        loc=None,\n",
    "        fontsize=12,\n",
    "        frameon=False\n",
    "    )\n",
    "    plt.gca().add_artist(plt.legend(\n",
    "        title=\"\",\n",
    "        loc=None,\n",
    "        fontsize=12,\n",
    "        frameon=False,\n",
    "    ))\n",
    "    plt.tight_layout()\n",
    "\n",
    "    # Display the plot\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "from sklearn.metrics import calinski_harabasz_score, silhouette_score, f1_score\n",
    "def compute_clustering_scores(\n",
    "    adata: sc.AnnData, cell_type_column: str = \"celltype_major\", use_pca: bool = True\n",
    ") -> Tuple[float, float]:\n",
    "    \"\"\"Compute the Calinski-Harabasz and Silhouette scores for an AnnData object.\"\"\"\n",
    "    if cell_type_column not in adata.obs:\n",
    "        raise ValueError(f\"Column '{cell_type_column}' must be present in adata.obs.\")\n",
    "\n",
    "    # Randomly sample a subset for efficiency (e.g., 10,000 cells)\n",
    "    n_cells = min(adata.n_obs, 10000)  # Limit to 10k cells\n",
    "    cell_indices = np.random.choice(adata.n_obs, n_cells, replace=False)\n",
    "\n",
    "    # Select features\n",
    "    features = adata.obsm[\"X_pca\"] if use_pca else adata.X\n",
    "    features = features[cell_indices, :]\n",
    "    labels = adata[cell_indices, :].obs[cell_type_column].values\n",
    "\n",
    "    # Compute scores\n",
    "    ch_score = calinski_harabasz_score(features, labels)\n",
    "    sh_score = silhouette_score(features, labels)\n",
    "    return ch_score, sh_score\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def preprocess_adata(\n",
    "    reference_adata: ad.AnnData, query_adata: ad.AnnData, transfer_column: str, min_counts: int = 5, max_counts: int = 500\n",
    ") -> ad.AnnData:\n",
    "    \"\"\"Annotate query AnnData object using a scRNA-seq reference atlas.\n",
    "\n",
    "    Args:\n",
    "    - reference_adata: ad.AnnData\n",
    "        Reference AnnData object containing the scRNA-seq atlas data.\n",
    "    - query_adata: ad.AnnData\n",
    "        Query AnnData object containing the data to be annotated.\n",
    "    - transfer_column: str\n",
    "        The name of the column in the reference atlas's `obs` to transfer to the query dataset.\n",
    "\n",
    "    Returns:\n",
    "    - query_adata: ad.AnnData\n",
    "        Annotated query AnnData object with transferred labels and UMAP coordinates from the reference.\n",
    "    \"\"\"\n",
    "    common_genes = list(set(reference_adata.var_names) & set(query_adata.var_names))\n",
    "    reference_adata = reference_adata[:, common_genes]\n",
    "    query_adata = query_adata[:, common_genes]\n",
    "    # min_counts = np.percentile(query_adata.obs.transcripts.values, 5)\n",
    "    # print(min_counts)\n",
    "    # max_counts = np.percentile(query_adata.obs.transcripts.values,99)\n",
    "    # print(max_counts)\n",
    "    sc.pp.filter_cells(query_adata, min_counts=min_counts)\n",
    "    sc.pp.filter_cells(query_adata, max_counts=max_counts)\n",
    "    query_adata.layers[\"raw\"] = query_adata.raw.X if query_adata.raw else query_adata.X\n",
    "    # query_adata.var[\"raw_counts\"] = query_adata.layers[\"raw\"].sum(axis=0)\n",
    "    sc.pp.normalize_total(query_adata, target_sum=1e4)\n",
    "    sc.pp.log1p(query_adata)\n",
    "    sc.tl.ingest(query_adata, reference_adata, obs=transfer_column, embedding_method='umap')\n",
    "    query_adata.obsm[\"X_umap_ref\"] = query_adata.obsm[\"X_umap\"]\n",
    "    sc.pp.pca(query_adata)\n",
    "    sc.pp.neighbors(query_adata)\n",
    "    sc.tl.umap(query_adata)\n",
    "    return query_adata\n",
    "\n",
    "\n",
    "def knn_adjacency_matrix(data_matrix, k):\n",
    "    # Construct a KDTree from the data matrix\n",
    "    tree = KDTree(data_matrix)\n",
    "\n",
    "    # Query the KDTree to find k nearest neighbors for each point\n",
    "    distances, indices = tree.query(data_matrix, k+1)  # k+1 because the nearest neighbor is the point itself\n",
    "\n",
    "    # Initialize a sparse matrix to hold the adjacency matrix\n",
    "    n = len(data_matrix)\n",
    "    adjacency_matrix = lil_matrix((n, n), dtype=int)\n",
    "\n",
    "    # Populate the adjacency matrix\n",
    "    for i in range(n):\n",
    "        for j in indices[i, 1:]:  # Exclude the first neighbor (itself)\n",
    "            adjacency_matrix[i, j] = 1\n",
    "            adjacency_matrix[j, i] = 1  # Assuming undirected graph\n",
    "\n",
    "    return adjacency_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "QvhltV87agrI"
   },
   "source": [
    "### Static values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "ztYoyox0agrI"
   },
   "outputs": [],
   "source": [
    "major_colors = {\n",
    "    \"Malignant\": \"#8E1600\",      # Deep muted red\n",
    "    \"T\": \"#1F77B4\",             # Muted blue\n",
    "    \"Plasma\": \"#76A7CB\",        # Soft blue\n",
    "    \"Myeloid\": \"#4C72B0\",       # Balanced deep blue\n",
    "    \"Fibroblast\": \"#A1B56C\",    # Muted green\n",
    "    \"B\": \"#64B5CD\",             # Soft cyan\n",
    "    \"Unknown\": \"#999999\",       # Neutral gray\n",
    "    \"Endothelial\": \"#5B8A72\",   # Earthy green\n",
    "    \"myofibroblast\": \"#C4A484\", # Muted brownish beige\n",
    "    \"Pericyte\": \"#E3C16F\",      # Warm sand\n",
    "    \"NK\": \"#6B4C9A\"             # Muted violet\n",
    "}\n",
    "\n",
    "xenium_tag = 'Xenium_V1_Human_Colon_Cancer_P1_CRC_Add_on_FFPE_outs'\n",
    "\n",
    "mtx_f = os.path.join(datadir, xenium_tag, 'cell_feature_matrix.h5')\n",
    "obs_f = os.path.join(datadir, xenium_tag, 'cells.csv.gz')\n",
    "sc_f  = os.path.join(datadir, 'CRC_UNB_10X_GSE132465.h5ad')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "0f0ySzplagrJ"
   },
   "source": [
    "## Spatial transcriptomics from the lens of single cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "dcizLZsGagrJ"
   },
   "source": [
    "### Read in the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that on both datasets, `sc.pp.pca`, `sc.pp.neighbors`, and `sc.tl.umap` was already applied. Further, both datasets were subsampled and prepared for UMAP plotting: \n",
    "```python\n",
    "# Xenium data\n",
    "xenium_data     = sc.read_10x_h5(mtx_f)\n",
    "obs             = pd.read_csv(obs_f, compression='gzip')\n",
    "xenium_data.obs = obs\n",
    "sc.pp.subsample(xenium_data, 0.2) #subset the xenium to smaller set\n",
    "\n",
    "# scRNAseq data\n",
    "scRNAseq_data   = sc.read(os.path.join(datadir, 'CRC_UNB_10X_GSE132465.h5ad'))\n",
    "sc.pp.subsample(scRNAseq_data, 0.2) #subset the xenium to smaller set\n",
    "\n",
    "sc.pp.pca(scRNAseq_data)\n",
    "sc.pp.neighbors(scRNAseq_data)\n",
    "sc.tl.umap(scRNAseq_data)\n",
    "\n",
    "sc.pp.pca(xenium_data)\n",
    "sc.pp.neighbors(xenium_data)\n",
    "sc.tl.leiden(xenium_data, 0.5)\n",
    "sc.tl.umap(xenium_data)\n",
    "\n",
    "# Subset to same genes\n",
    "genes           = list(set(xenium_data.var_names) & set(scRNAseq_data.var_names))\n",
    "xenium_data     = xenium_data[:, genes]\n",
    "scRNAseq_data   = scRNAseq_data[:, genes]\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xenium_data = sc.read_h5ad(os.path.join(datadir, 'xenium_data.h5ad'))\n",
    "scRNAseq_data = sc.read_h5ad(os.path.join(datadir, 'scRNAseq_data.h5ad'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "aq3dnz-lcxvq"
   },
   "outputs": [],
   "source": [
    "# Subset to same genes\n",
    "genes           = list(set(xenium_data.var_names) & set(scRNAseq_data.var_names))\n",
    "xenium_data     = xenium_data[:, genes]\n",
    "scRNAseq_data   = scRNAseq_data[:, genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "XDOZdXmiagrK",
    "outputId": "dac84c26-50d4-4da6-eb9b-f318e277ab4f"
   },
   "outputs": [],
   "source": [
    "print(scRNAseq_data)\n",
    "print(xenium_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "H10ONdTCagrL"
   },
   "outputs": [],
   "source": [
    "xenium_data.obsm['spatial'] = xenium_data.obs[['x_centroid', 'y_centroid']].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "SAX2V7NMTW2L"
   },
   "source": [
    "\n",
    "\n",
    "### Good old UMAPs!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 883
    },
    "id": "gK3cgtoC2In9",
    "outputId": "306c2e23-dabc-4138-e84d-d2bd7716a9c9"
   },
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(1, 2, figsize=(8, 4))\n",
    "sc.pl.umap(\n",
    "        xenium_data,\n",
    "        color='leiden',\n",
    "        show=False,\n",
    "        ax=axs[0],\n",
    "        add_outline=True,\n",
    "        legend_loc=\"on data\",\n",
    "        # palette=major_colors,\n",
    "        title='Xenium',\n",
    "    )\n",
    "sc.pl.umap(\n",
    "        scRNAseq_data,\n",
    "        color='annotation_major',\n",
    "        show=False,\n",
    "        ax = axs[1],\n",
    "        legend_loc=\"on data\",\n",
    "        add_outline=True,\n",
    "        palette=major_colors,\n",
    "        title='scRNAseq',\n",
    "    )\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "OWpuRS9VagrM",
    "outputId": "3b281b02-1b85-43d6-bb46-1c6a4037adef"
   },
   "outputs": [],
   "source": [
    "from matplotlib.pyplot import rc_context\n",
    "with rc_context({\"figure.figsize\": (4, 4)}):\n",
    "    sc.pl.spatial(\n",
    "            xenium_data,\n",
    "            color=[\n",
    "                'leiden'\n",
    "            ],\n",
    "            # library_id=\"spatial\",\n",
    "            # show=False,\n",
    "            # ax = axs[1],\n",
    "            ncols=2,\n",
    "            spot_size=25,\n",
    "            title='Xenium',\n",
    "            # save='leiden_space.pdf'\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "FlwrAG8Yv5dv"
   },
   "source": [
    "\n",
    "\n",
    "## Preprocessing (as in scRNAseq!)\n",
    "\n",
    "### 1. Quality control and Filtering\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "LnkGt5Tbwwam"
   },
   "outputs": [],
   "source": [
    "sc.pp.calculate_qc_metrics(xenium_data, percent_top=(10, 20, 50, 150), inplace=True)\n",
    "#List of ranks (where genes are ranked by expression) at which the cumulative proportion of expression will be reported as a percentage.\n",
    "#This can be used to assess library complexity.\n",
    "\n",
    "#e.g. percent_top=[50] : for each cell cumulative expression of 50 highest expressed genes from total expression is calculated\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "JCaG2HLdwzey"
   },
   "outputs": [],
   "source": [
    "cprobes = (\n",
    "    xenium_data.obs[\"control_probe_counts\"].sum() / xenium_data.obs[\"total_counts\"].sum() * 100\n",
    ") # pct of control probes of all probes -> used to control for\n",
    "cwords = (\n",
    "    xenium_data.obs[\"control_codeword_counts\"].sum() / xenium_data.obs[\"total_counts\"].sum() * 100\n",
    ") # pct of control codewords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 427
    },
    "id": "0nCEYF1Zw8sp",
    "outputId": "547cdcde-5e5c-4b30-b92a-28c6049b6e56"
   },
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(1, 4, figsize=(8, 2))\n",
    "\n",
    "axs[0].set_title(\"Total transcripts per cell\")\n",
    "sns.histplot(\n",
    "    xenium_data.obs[\"total_counts\"],\n",
    "    kde=False,\n",
    "    ax=axs[0],\n",
    ") #plot histogram of total number of transripts per cell\n",
    "\n",
    "axs[1].set_title(\"Unique transcripts per cell\")\n",
    "sns.histplot(\n",
    "    xenium_data.obs[\"n_genes_by_counts\"],\n",
    "    kde=False,\n",
    "    ax=axs[1],\n",
    ") #plot histogram of unique number of transripts per cell\n",
    "\n",
    "\n",
    "axs[2].set_title(\"Area of segmented cells\")\n",
    "sns.histplot(\n",
    "    xenium_data.obs[\"cell_area\"],\n",
    "    kde=False,\n",
    "    ax=axs[2],\n",
    ") #plot histogram of cell sizez/areas\n",
    "\n",
    "axs[3].set_title(\"Nucleus ratio\")\n",
    "sns.histplot(\n",
    "    xenium_data.obs[\"nucleus_area\"] / xenium_data.obs[\"cell_area\"],\n",
    "    kde=False,\n",
    "    ax=axs[3],\n",
    ") #plot histogram of ratios of nuclei-to-cell sizes (-areas)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "xwXRt02c9tyH"
   },
   "outputs": [],
   "source": [
    "sc.pp.filter_cells(xenium_data, min_counts = 5)\n",
    "sc.pp.filter_genes(xenium_data, min_cells = 100)\n",
    "#filter out cells with less than 5 transcripts detected and genes that are detected in less than 100 cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "QjhAmWrcagrN",
    "outputId": "7d016375-c3b4-4131-c2b3-016bdd049e38"
   },
   "outputs": [],
   "source": [
    "from matplotlib.pyplot import rc_context\n",
    "with rc_context({\"figure.figsize\": (4, 4)}):\n",
    "    sc.pl.spatial(\n",
    "            xenium_data,\n",
    "            color=[\n",
    "                \"total_counts\",\n",
    "                \"n_genes_by_counts\",\n",
    "                \"cell_area\",\n",
    "                'leiden'\n",
    "            ],\n",
    "            # library_id=\"spatial\",\n",
    "            # show=False,\n",
    "            # ax = axs[1],\n",
    "            ncols=2,\n",
    "            spot_size=20,\n",
    "            # title='Xenium',\n",
    "            # save='metrics.pdf'\n",
    "        )\n",
    "    #overlay 2D sample-maps with:\n",
    "      #total counts of transcripts per cell\n",
    "      #number of unique transcripts per cell"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "2ltUeqDeN26J"
   },
   "source": [
    "### 2. Normalization and dimension reduction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `xenium_data` has log-normalized data in `adata.X` and raw counts in `adata.layers[\"counts\"]`\n",
    "```python \n",
    "xenium_data.layers[\"counts\"] = xenium_data.X.copy() # backup of raw counts\n",
    "sc.pp.normalize_total(xenium_data, inplace=True) # normalizes each cells total gene count to the same total count -> highly expressed cells should not be overrepresented\n",
    "sc.pp.log1p(xenium_data) # Applies a log(1 + x) transformation to scale down gene expression counts to lower values\n",
    "sc.pp.pca(xenium_data) #Reduction of dimensions to the principle components that explain most of the variance\n",
    "sc.pp.neighbors(xenium_data) # constructs a K-Nearest Neighbors (KNN) graph using the PCA-reduced data ->  umap (and leiden clustering) require this neighborhood graph\n",
    "sc.tl.umap(xenium_data) # Umap projects cells into a lower-dimensional space (typically 2D or 3D) for visualization ->\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "5zIqpIXfmKVe",
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "xenium_data = sc.read_h5ad(os.path.join(datadir, 'xenium_data_mapped.h5ad'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "n7fNGuA0agrN",
    "outputId": "06032fc3-b06d-470a-f0b4-2700e0250b1a"
   },
   "outputs": [],
   "source": [
    "from matplotlib.pyplot import rc_context\n",
    "with rc_context({\"figure.figsize\": (4, 4)}): # --> 4x4 inches image\n",
    "    sc.pl.umap(\n",
    "            xenium_data, # the plot uses data stored in the umap output .obsm of the ann data file xenium data\n",
    "            color=[\n",
    "                \"total_counts\",\n",
    "                \"n_genes_by_counts\",\n",
    "                \"cell_area\",\n",
    "                'leiden'\n",
    "            ],\n",
    "            # library_id=\"spatial\",\n",
    "            # show=False,\n",
    "            # ax = axs[1],\n",
    "            ncols=2,\n",
    "            # spot_size=10,\n",
    "            title='Xenium',\n",
    "            # save='metrics_umap.pdf'\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "B9eqi8mzN9Wa"
   },
   "source": [
    "## Cell type annotation (as in scRNAseq!)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "a8MzdMh-T5Vi"
   },
   "source": [
    "### 3. Label transfer from a reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset to same genes again\n",
    "genes           = list(set(xenium_data.var_names) & set(scRNAseq_data.var_names))\n",
    "xenium_data     = xenium_data[:, genes]\n",
    "scRNAseq_data   = scRNAseq_data[:, genes]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This takes a while so we precomputed: \n",
    "```python\n",
    "sc.tl.ingest(xenium_data, scRNAseq_data, obs='annotation_major') #maps new cells (Xenium spatial transcriptomics data) to scRNA-seq reference dataset based on shared transcriptomic profiles\n",
    "xenium_data.uns['annotation_major_colors'] = scRNAseq_data.uns['annotation_major_colors']\n",
    "xenium_data.obsm['umap_ingest'] = xenium_data.obsm['X_umap']\n",
    "sc.tl.umap(xenium_data)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scRNAseq_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "e4So1EgMagrO",
    "outputId": "ce4b38e8-0097-4328-a2f4-3c7c6c4c8637"
   },
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(1, 2, figsize=(6, 2))\n",
    "\n",
    "labels = scRNAseq_data.obs['annotation_major'].value_counts().index\n",
    "sns.countplot(\n",
    "    scRNAseq_data.obs, x=\"annotation_major\", #hue='annotation_major',\n",
    "    # bins=50,\n",
    "    # hue_order=order,\n",
    "    palette=major_colors,\n",
    "    order = labels,\n",
    "    ax=axs[0],\n",
    "    # gap=.25\n",
    ")\n",
    "axs[0].set_ylabel('# Cells', fontsize=7.5)\n",
    "axs[0].set_title('scRNAseq')\n",
    "axs[0].set_xticklabels(labels= labels, rotation=45, ha='right')\n",
    "\n",
    "\n",
    "labels = xenium_data.obs['annotation_major'].value_counts().index\n",
    "sns.countplot(\n",
    "    xenium_data.obs, x=\"annotation_major\", #hue='annotation_major',\n",
    "    # bins=50,\n",
    "    # hue_order=order,\n",
    "    palette=major_colors,\n",
    "    order = labels,\n",
    "    ax=axs[1],\n",
    "    # gap=.25\n",
    ")\n",
    "axs[1].set_ylabel('# Cells', fontsize=7.5)\n",
    "axs[1].set_title('Xenium')\n",
    "# axs[0].set(ylim=(0, 25_000))\n",
    "axs[1].set_xticklabels(labels= labels, rotation=45, ha='right')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 349
    },
    "id": "hgBoV9CwEGLA",
    "outputId": "e00a3092-4da1-4d5f-e0d6-43f1e529c8f5"
   },
   "outputs": [],
   "source": [
    "with rc_context({\"figure.figsize\": (3, 6)}):\n",
    "    sc.pl.spatial(\n",
    "            xenium_data,\n",
    "            color=[\n",
    "                \"annotation_major\",\n",
    "                'leiden'\n",
    "            ],\n",
    "            # palette={major_colors, None],\n",
    "            # library_id=\"spatial\",\n",
    "            # show=False,\n",
    "            # ax = axs[1],\n",
    "            ncols=2,\n",
    "            spot_size=20,\n",
    "            # title='Xenium',\n",
    "            # save='metrics.pdf'\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "afV1xVlSagrP",
    "outputId": "19ee1ce1-1176-4ba4-81bd-7a0f59f6693f"
   },
   "outputs": [],
   "source": [
    "xenium_data.uns['annotation_major_colors'] = scRNAseq_data.uns['annotation_major_colors']\n",
    "\n",
    "fig, axs = plt.subplots(1, 3, figsize=(10, 3))\n",
    "sc.pl.umap(\n",
    "        xenium_data,\n",
    "        color='annotation_major',\n",
    "        show=False,\n",
    "        ax=axs[0],\n",
    "        add_outline=True,\n",
    "        legend_loc=\"on data\",\n",
    "        # palette=major_colors,\n",
    "        title='Xenium'\n",
    "    )\n",
    "\n",
    "sc.pl.embedding(\n",
    "        xenium_data,\n",
    "        color='annotation_major',\n",
    "        show=False,\n",
    "        ax=axs[1],\n",
    "        add_outline=True,\n",
    "        legend_loc=\"on data\",\n",
    "        basis='umap_ingest',\n",
    "        # palette=major_colors,\n",
    "        title='Xenium'\n",
    "    )\n",
    "\n",
    "sc.pl.umap(\n",
    "        scRNAseq_data,\n",
    "        color='annotation_major',\n",
    "        show=False,\n",
    "        ax = axs[2],\n",
    "        legend_loc=\"on data\",\n",
    "        add_outline=True,\n",
    "        palette=major_colors,\n",
    "        title='scRNAseq'\n",
    "    )\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "RB4h0g6magrP"
   },
   "outputs": [],
   "source": [
    " # Drop rare cell types! (CAUTION!)\n",
    "min_cells_per_celltype = 50\n",
    "vals = xenium_data.obs.annotation_major.value_counts().values > min_cells_per_celltype\n",
    "labels = xenium_data.obs.annotation_major.value_counts().index\n",
    "types = [i for indx,i in enumerate(labels) if vals[indx] == True]\n",
    "xenium_data = xenium_data[xenium_data.obs.annotation_major.isin(types)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 972
    },
    "id": "MQcw3FbTKrmP",
    "outputId": "3e0452c1-3f4f-4c2f-d84f-c020f2a1ad7e"
   },
   "outputs": [],
   "source": [
    "with rc_context({\"figure.figsize\": (3, 3)}):\n",
    "    xenium_data.obs['annotation_major'] = xenium_data.obs['annotation_major'].cat.remove_unused_categories()\n",
    "    cmtx = sc.metrics.confusion_matrix(\"leiden\", \"annotation_major\", xenium_data.obs)\n",
    "    sns.heatmap(cmtx)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "eM4M2TvdT8t8"
   },
   "source": [
    "### 4. Exploring marker expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "H3yYD7ByagrQ"
   },
   "outputs": [],
   "source": [
    "xenium_data.obs['annotation_major'] = xenium_data.obs['annotation_major'].cat.remove_unused_categories()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "id": "8OaNbsPThKUz",
    "outputId": "6de6a73d-6790-4aba-ebea-032d38034525"
   },
   "outputs": [],
   "source": [
    "with rc_context({\"figure.figsize\": (4, 2)}):\n",
    "    sc.tl.rank_genes_groups(scRNAseq_data, 'annotation_major', method='logreg',max_iter=100)\n",
    "    sc.pl.rank_genes_groups_dotplot(scRNAseq_data, n_genes=4, standard_scale='var', title='scRNAseq')\n",
    "    sc.tl.rank_genes_groups(xenium_data, 'annotation_major', method='logreg',max_iter=100)\n",
    "    sc.pl.rank_genes_groups_dotplot(xenium_data, n_genes=4, standard_scale='var', title='Xenium')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "2lD1rcaKORRG"
   },
   "source": [
    "## Bring in the Space!\n",
    "\n",
    "[Let's look at the data!](https://xenium.10xgenomics.com/?image=s3%2F10x.files%2Fxenium%2Fpreview%2Fhbreast%2Fexperiment.xenium&zoom=0.0217&target=19259.42_12511.18&z=17&off=&layers=cell~image&mi_v=true&nav=true&axes=false&feature=icon&feature_psc=6&feature_ps=circles&ii_n=H%26E%20Image&ii_p=s3%2F10x.files%2Fxenium%2Fpreview%2FXenium_FFPE_Human_Breast_Cancer_Rep1%2FXenium_FFPE_Human_Breast_Cancer_Rep1_he_image_registered.ome.tif&ii_t=AAAAAAAA8D8AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADwPwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAPA%2F&ii_v=true&feature_off=313_541&bin=10&d_c=Viridis&d_o=0.8&cell_bt=cell&cell_c=groups&cell_g=0&cell_o=0.5&ct_c=Viridis&ct_o=0.8&cell_v=filled)\n",
    "\n",
    "### 5. Neighborhood graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "XvZ2gE-QYEXT"
   },
   "outputs": [],
   "source": [
    "sq.gr.spatial_neighbors(xenium_data, coord_type=\"generic\", n_neighs=15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "MiBKjxH9agrV",
    "outputId": "47893f60-409d-43d9-fedc-a965103352d9"
   },
   "outputs": [],
   "source": [
    "xenium_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 837
    },
    "id": "XxDTW8B-YMn8",
    "outputId": "00c9dc8a-56c2-488d-a3bc-a8b0e483adf3"
   },
   "outputs": [],
   "source": [
    "sc.tl.paga(xenium_data, groups='annotation_major', neighbors_key=\"spatial_neighbors\", copy=False)\n",
    "sc.pl.paga(xenium_data, max_edge_width=2, node_size_scale=10, threshold= .8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 576,
     "referenced_widgets": [
      "89e3ee2d8696458680a179fdf72917ab",
      "8bc2999074f24618870722e84623a814",
      "66e03e4fe0a14c6aab1241185c1f7dfe",
      "e5b16d0b9e614f158647f7a325d958ce",
      "fb85f0187447442f86c5610ecca889f2",
      "9d07fefeed824cdebad000957a575954",
      "923ffba495834b1caff65f16c98d2aa0",
      "6f8f06a6704a4f25be43adc7bc9ff987",
      "db616d51e2d04a89aef0a4cf54d1d96c",
      "68bbfd0ac0d54ceeb4d5f2e29e65908d",
      "52abc750e0034b04bfddb229d1d7c883",
      "a0dd57e260fa463b85cdecc00c658564"
     ]
    },
    "id": "xOoEKIUmEpp3",
    "outputId": "6b105d17-96d2-4333-cc70-16367eee6fd7"
   },
   "source": [
    "```python\n",
    "sq.gr.nhood_enrichment(xenium_data, cluster_key=\"annotation_major\")\n",
    "```\n",
    "\n",
    "This added \n",
    "\n",
    "    - :attr:`anndata.AnnData.uns` ``['{cluster_key}_nhood_enrichment']['zscore']`` - the enrichment z-score.\n",
    "    - :attr:`anndata.AnnData.uns` ``['{cluster_key}_nhood_enrichment']['count']`` - the enrichment count."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "Gptxjq3MagrW",
    "outputId": "665e0475-8f7b-41cd-c581-137c383d84cc"
   },
   "outputs": [],
   "source": [
    "xenium_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "ptnvea3kagrW",
    "outputId": "e84d77a3-35fc-41d6-b883-5d57792830cd"
   },
   "outputs": [],
   "source": [
    "sq.pl.nhood_enrichment(\n",
    "    xenium_data,\n",
    "    cluster_key=\"annotation_major\",\n",
    "    method=\"average\",\n",
    "    cmap=\"inferno\",\n",
    "    vmin=-50,\n",
    "    vmax=100,\n",
    "    figsize=(5, 5),\n",
    "    title='celltype colocalization'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "oGy94hKNOb9B"
   },
   "source": [
    "### 6. Spatially informative genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 716
    },
    "id": "vmwKiFlbFZjg",
    "outputId": "db2bba8c-97f5-4732-e784-94822e051cc7"
   },
   "outputs": [],
   "source": [
    "with rc_context({\"figure.figsize\": (2, 2)}):\n",
    "    sq.gr.spatial_autocorr(xenium_data, mode=\"moran\")\n",
    "    top_genes = xenium_data.uns[\"moranI\"].head().index[:3]\n",
    "    low_genes = xenium_data.uns[\"moranI\"].tail().index[:3]\n",
    "    sq.pl.spatial_scatter(\n",
    "        xenium_data, shape=None, color=top_genes.append(low_genes), size=3, ncols=3, legend_loc=None,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "BgOY4WfmUKvP"
   },
   "source": [
    "### 7. Niche detection\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "wreRXh2i-6US"
   },
   "source": [
    "Idea: locate spatial niches by clustering cells based on proportion of surrounding celltypes\n",
    "\n",
    "1. encode cell type with one-hot encoding (celltype of cell: 1, other celltypes:0)\n",
    "2. create weighted sum of celltypes in spatial proximity & normalise\n",
    "3. compute K-nearest neighbour graph based on this representation -> cells with similar proportion of surrounding celltypes are connected\n",
    "4. compute leiden clustering on KNN graph to identify clusters (= niches)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "Urv2SvuUMIsy",
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from sklearn.neighbors import NearestNeighbors\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "ohe = OneHotEncoder()\n",
    "\n",
    "xenium_data2 = xenium_data.copy()\n",
    "xenium_data2.X = xenium_data2.X.toarray()\n",
    "sc.pp.subsample(xenium_data2, 0.1)\n",
    "\n",
    "celltype_one = ohe.fit_transform(xenium_data2.obs[['annotation_major']])\n",
    "celltype_dist = xenium_data2.obsp['spatial_connectivities'] @ celltype_one\n",
    "celltype_dist /= celltype_dist.sum(axis=1)\n",
    "xenium_data2.obsm['celltype_dist'] = celltype_dist.tocsc()\n",
    "# sc.pp.subsample(xenium_data2, 0.1)\n",
    "a = knn_adjacency_matrix(xenium_data2.obsm['celltype_dist'].toarray(), k = 100)\n",
    "xenium_data2.obsp['niche_connectivities_minor'] = a.toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.leiden(xenium_data2, adjacency=xenium_data2.obsp['niche_connectivities_minor'], key_added='niche', resolution=0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 486
    },
    "id": "H5ePh-PeCTtE",
    "outputId": "951eef83-67e9-4625-9009-c7cca3f2eb0d"
   },
   "outputs": [],
   "source": [
    "sq.pl.spatial_scatter(\n",
    "    xenium_data2,\n",
    "    library_id=\"spatial\",\n",
    "    shape=None,\n",
    "    color=[\n",
    "        \"niche\",\n",
    "        \"annotation_major\"\n",
    "    ],\n",
    "    wspace=0.4,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 867
    },
    "id": "dpQqirz3Hh3V",
    "outputId": "1894cac8-1f3b-4d1f-b806-dfbbf0211d47"
   },
   "outputs": [],
   "source": [
    "cmtx = sc.metrics.confusion_matrix(\"annotation_major\", \"niche\", xenium_data2.obs)\n",
    "sns.heatmap(cmtx)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "nJ2Cwf7TSzx6"
   },
   "source": [
    "## Extra links\n",
    "\n",
    "* [`SpatialData`, a framework for wrangling one and multiple spatial omics datasets across samples and technologies](https://spatialdata.scverse.org/en/latest/)\n",
    "* [Celltype annotation for visium using cell2location](https://cell2location.readthedocs.io/en/latest/)\n",
    "* [Linear alignment of Xenium and Visium (10X general)](https://www.10xgenomics.com/support/software/xenium-explorer/latest/tutorials/xe-image-alignment)\n",
    "* [None-linear alignment of Xenium and Visium (STalign)](https://jef.works/blog/2023/12/27/alignment-of-spatial-transcriptomics-data-using-STalign/)\n",
    "* [Xenium Cell Segmentation; best practices](https://www.biorxiv.org/content/10.1101/2023.02.13.528102v1)"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [
    "B5zAOlJWFvqF",
    "VRg_B1OWagrF",
    "dcizLZsGagrJ",
    "SAX2V7NMTW2L"
   ],
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python (01 Spatial)",
   "language": "python",
   "name": "01-spatial"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "52abc750e0034b04bfddb229d1d7c883": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "66e03e4fe0a14c6aab1241185c1f7dfe": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "FloatProgressModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "FloatProgressModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "ProgressView",
      "bar_style": "success",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_6f8f06a6704a4f25be43adc7bc9ff987",
      "max": 1000,
      "min": 0,
      "orientation": "horizontal",
      "style": "IPY_MODEL_db616d51e2d04a89aef0a4cf54d1d96c",
      "value": 1000
     }
    },
    "68bbfd0ac0d54ceeb4d5f2e29e65908d": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "6f8f06a6704a4f25be43adc7bc9ff987": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "89e3ee2d8696458680a179fdf72917ab": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HBoxModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HBoxModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HBoxView",
      "box_style": "",
      "children": [
       "IPY_MODEL_8bc2999074f24618870722e84623a814",
       "IPY_MODEL_66e03e4fe0a14c6aab1241185c1f7dfe",
       "IPY_MODEL_e5b16d0b9e614f158647f7a325d958ce"
      ],
      "layout": "IPY_MODEL_fb85f0187447442f86c5610ecca889f2"
     }
    },
    "8bc2999074f24618870722e84623a814": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_9d07fefeed824cdebad000957a575954",
      "placeholder": "",
      "style": "IPY_MODEL_923ffba495834b1caff65f16c98d2aa0",
      "value": "100%"
     }
    },
    "923ffba495834b1caff65f16c98d2aa0": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "9d07fefeed824cdebad000957a575954": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "db616d51e2d04a89aef0a4cf54d1d96c": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "ProgressStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "ProgressStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "bar_color": null,
      "description_width": ""
     }
    },
    "e5b16d0b9e614f158647f7a325d958ce": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_68bbfd0ac0d54ceeb4d5f2e29e65908d",
      "placeholder": "",
      "style": "IPY_MODEL_52abc750e0034b04bfddb229d1d7c883",
      "value": "1000/1000[00:03&lt;00:00,312.77/s]"
     }
    },
    "fb85f0187447442f86c5610ecca889f2": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
